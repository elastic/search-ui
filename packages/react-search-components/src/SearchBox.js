import PropTypes from "prop-types";
import React from "react";

function SearchBox(props) {
  const { isFocused, inputProps, onChange, onSubmit, value } = props;
  const focusedClass = isFocused ? "focus" : "";

  return (
    <form className="sui-search-box" onSubmit={onSubmit}>
      <input
        className={`sui-search-box__text-input ${focusedClass}`}
        onChange={onChange}
        type="text"
        value={value}
        placeholder="Search your documents&#8230;"
        {...inputProps}
      />
      <input
        type="submit"
        value="Search"
        className="button sui-search-box__submit"
      />
    </form>
  );
}

SearchBox.propTypes = {
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  inputProps: PropTypes.object,
  isFocused: PropTypes.bool
};

export default SearchBox;
