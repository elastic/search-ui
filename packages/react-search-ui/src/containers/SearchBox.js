import PropTypes from "prop-types";
import { Component } from "react";
import { SearchBox } from "@elastic/react-search-ui-views";

import { withSearch } from "..";
import { Result, Suggestion } from "../types";

export class SearchBoxContainer extends Component {
  static propTypes = {
    // Props
    autocompleteMinimumCharacters: PropTypes.number,
    autocompleteResults: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.shape({
        clickThroughTags: PropTypes.arrayOf(PropTypes.string),
        linkTarget: PropTypes.string,
        sectionTitle: PropTypes.string,
        shouldTrackClickThrough: PropTypes.bool,
        titleField: PropTypes.string.isRequired,
        urlField: PropTypes.string.isRequired
      })
    ]),
    autocompleteSuggestions: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.exact({
        sectionTitle: PropTypes.string
      }),
      PropTypes.objectOf(
        PropTypes.exact({
          sectionTitle: PropTypes.string
        })
      )
    ]),
    autocompleteView: PropTypes.func,
    className: PropTypes.string,
    shouldClearFilters: PropTypes.bool,
    debounceLength: PropTypes.number,
    inputProps: PropTypes.object,
    inputView: PropTypes.func,
    onSelectAutocomplete: PropTypes.func,
    onSubmit: PropTypes.func,
    searchAsYouType: PropTypes.bool,
    view: PropTypes.func,
    // State
    autocompletedResults: PropTypes.arrayOf(Result).isRequired,
    autocompletedSuggestions: PropTypes.objectOf(PropTypes.arrayOf(Suggestion))
      .isRequired,
    searchTerm: PropTypes.string.isRequired,
    // Actions
    setSearchTerm: PropTypes.func.isRequired,
    trackAutocompleteClickThrough: PropTypes.func.isRequired
  };

  static defaultProps = {
    autocompleteMinimumCharacters: 0,
    shouldClearFilters: true
  };

  state = {
    isFocused: false
  };

  handleFocus = () => {
    this.setState({
      isFocused: true
    });
  };

  handleBlur = () => {
    this.setState({
      isFocused: false
    });
  };

  completeSuggestion = searchTerm => {
    const { shouldClearFilters, setSearchTerm } = this.props;
    setSearchTerm(searchTerm, {
      shouldClearFilters
    });
  };

  handleSubmit = e => {
    const { shouldClearFilters, searchTerm, setSearchTerm } = this.props;

    e.preventDefault();
    setSearchTerm(searchTerm, {
      shouldClearFilters
    });
  };

  handleChange = value => {
    const {
      autocompleteMinimumCharacters,
      autocompleteResults,
      autocompleteSuggestions,
      shouldClearFilters,
      searchAsYouType,
      setSearchTerm,
      debounceLength
    } = this.props;

    const options = {
      autocompleteMinimumCharacters,
      ...((autocompleteResults ||
        autocompleteSuggestions ||
        searchAsYouType) && {
        debounce: debounceLength || 200
      }),
      shouldClearFilters,
      refresh: !!searchAsYouType,
      autocompleteResults: !!autocompleteResults,
      autocompleteSuggestions: !!autocompleteSuggestions
    };

    setSearchTerm(value, options);
  };

  handleNotifyAutocompleteSelected = selection => {
    const { autocompleteResults, trackAutocompleteClickThrough } = this.props;
    // Because suggestions don't count as clickthroughs, only
    // results
    if (
      autocompleteResults &&
      autocompleteResults.shouldTrackClickThrough !== false &&
      !selection.suggestion
    ) {
      const { clickThroughTags = [] } = autocompleteResults;
      const id = selection.id.raw;
      trackAutocompleteClickThrough(id, clickThroughTags);
    }
  };

  defaultOnSelectAutocomplete = selection => {
    if (!selection) return;

    const { autocompleteResults } = this.props;

    this.handleNotifyAutocompleteSelected(selection);
    if (!selection.suggestion) {
      const url = selection[autocompleteResults.urlField]
        ? selection[autocompleteResults.urlField].raw
        : "";
      if (url) {
        const target = autocompleteResults.linkTarget || "_self";
        window.open(url, target);
      }
    } else {
      this.completeSuggestion(selection.suggestion);
    }
  };

  render() {
    const { isFocused } = this.state;
    const {
      autocompleteMinimumCharacters,
      autocompleteResults,
      autocompleteSuggestions,
      autocompletedResults,
      autocompletedSuggestions,
      className,
      autocompleteView,
      inputProps,
      inputView,
      onSelectAutocomplete,
      onSubmit,
      searchTerm,
      view,
      ...rest
    } = this.props;

    const View = view || SearchBox;
    const useAutocomplete =
      (!!autocompleteResults || !!autocompleteSuggestions) &&
      searchTerm.length >= autocompleteMinimumCharacters;
    const autocompletedSuggestionsCount = Object.entries(
      autocompletedSuggestions
      // eslint-disable-next-line no-unused-vars
    ).reduce((acc, [_, value]) => acc + value.length, 0);
    const allAutocompletedItemsCount =
      autocompletedSuggestionsCount + autocompletedResults.length;

    let handleOnSelectAutocomplete;
    if (onSelectAutocomplete) {
      handleOnSelectAutocomplete = selection => {
        onSelectAutocomplete(
          selection,
          {
            notifyAutocompleteSelected: this.handleNotifyAutocompleteSelected,
            completeSuggestion: this.completeSuggestion,
            autocompleteResults: this.props.autocompleteResults
          },
          this.defaultOnSelectAutocomplete
        );
      };
    }

    return View({
      allAutocompletedItemsCount: allAutocompletedItemsCount,
      autocompleteView,
      autocompleteResults: autocompleteResults,
      autocompleteSuggestions: autocompleteSuggestions,
      autocompletedResults: autocompletedResults,
      autocompletedSuggestions: autocompletedSuggestions,
      className,
      autocompletedSuggestionsCount: autocompletedSuggestionsCount,
      completeSuggestion: this.completeSuggestion,
      isFocused: isFocused,
      notifyAutocompleteSelected: this.handleNotifyAutocompleteSelected,
      onChange: value => this.handleChange(value),
      onSelectAutocomplete:
        handleOnSelectAutocomplete || this.defaultOnSelectAutocomplete,
      onSubmit: onSubmit
        ? e => {
            e.preventDefault();
            onSubmit(searchTerm);
          }
        : this.handleSubmit,
      useAutocomplete: useAutocomplete,
      value: searchTerm,
      inputProps: {
        onFocus: this.handleFocus,
        onBlur: this.handleBlur,
        ...inputProps
      },
      inputView,
      ...rest
    });
  }
}

export default withSearch(
  ({
    autocompletedResults,
    autocompletedSuggestions,
    searchTerm,
    setSearchTerm,
    trackAutocompleteClickThrough
  }) => ({
    autocompletedResults,
    autocompletedSuggestions,
    searchTerm,
    setSearchTerm,
    trackAutocompleteClickThrough
  })
)(SearchBoxContainer);
