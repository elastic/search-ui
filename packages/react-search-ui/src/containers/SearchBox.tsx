import React, { Component } from "react";
import {
  SearchBox,
  SearchBoxContainerProps,
  SearchBoxContainerContext
} from "@elastic/react-search-ui-views";

import { withSearch } from "..";

export class SearchBoxContainer extends Component<SearchBoxContainerProps> {
  static defaultProps = {
    autocompleteMinimumCharacters: 0,
    shouldClearFilters: true
  };

  state = {
    isFocused: false
  };

  handleFocus = () => {
    this.setState({
      isFocused: true
    });
  };

  handleBlur = () => {
    this.setState({
      isFocused: false
    });
  };

  completeSuggestion = (searchTerm) => {
    const { shouldClearFilters, setSearchTerm } = this.props;
    setSearchTerm(searchTerm, {
      shouldClearFilters
    });
  };

  handleSubmit = (e) => {
    const { shouldClearFilters, searchTerm, setSearchTerm } = this.props;

    e.preventDefault();
    setSearchTerm(searchTerm, {
      shouldClearFilters
    });
  };

  handleChange = (value) => {
    const {
      autocompleteMinimumCharacters,
      autocompleteResults,
      autocompleteSuggestions,
      shouldClearFilters,
      searchAsYouType,
      setSearchTerm,
      debounceLength
    } = this.props;

    const options = {
      autocompleteMinimumCharacters,
      ...((autocompleteResults ||
        autocompleteSuggestions ||
        searchAsYouType) && {
        debounce: debounceLength || 200
      }),
      shouldClearFilters,
      refresh: !!searchAsYouType,
      autocompleteResults: !!autocompleteResults,
      autocompleteSuggestions: !!autocompleteSuggestions
    };

    setSearchTerm(value, options);
  };

  handleNotifyAutocompleteSelected = (selection) => {
    const {
      autocompleteResults,
      trackAutocompleteClickThrough,
      trackAutocompleteSuggestionClickThrough
    } = this.props;

    // results
    if (autocompleteResults) {
      const autocompleteResultsConfig =
        autocompleteResults === true
          ? { clickThroughTags: [], shouldTrackClickThrough: true }
          : autocompleteResults;

      if (
        !selection.suggestion &&
        autocompleteResultsConfig.shouldTrackClickThrough !== false
      ) {
        const { clickThroughTags = [] } = autocompleteResultsConfig;
        const id = selection.id?.raw;
        trackAutocompleteClickThrough(id, clickThroughTags);
      }

      if (selection.suggestion) {
        trackAutocompleteSuggestionClickThrough(
          selection.suggestion,
          selection.index,
          []
        );
      }
    }
  };

  defaultOnSelectAutocomplete = (selection) => {
    if (!selection) return;

    const { autocompleteResults } = this.props;

    this.handleNotifyAutocompleteSelected(selection);
    if (!selection.suggestion && typeof autocompleteResults !== "boolean") {
      const url = selection[autocompleteResults.urlField]
        ? selection[autocompleteResults.urlField].raw
        : "";
      if (url) {
        const target =
          (typeof autocompleteResults !== "boolean" &&
            autocompleteResults.linkTarget) ||
          "_self";
        window.open(url, target);
      }
    } else {
      this.completeSuggestion(selection.suggestion);
    }
  };

  render() {
    const { isFocused } = this.state;
    const {
      autocompleteMinimumCharacters,
      autocompleteResults,
      autocompleteSuggestions,
      autocompletedResults,
      autocompletedSuggestions,
      className,
      autocompleteView,
      inputProps,
      inputView,
      onSelectAutocomplete,
      onSubmit,
      searchTerm,
      view,
      ...rest
    } = this.props;

    const View = view || SearchBox;
    const useAutocomplete =
      (!!autocompleteResults || !!autocompleteSuggestions) &&
      searchTerm.length >= autocompleteMinimumCharacters;
    const autocompletedSuggestionsCount = Object.entries(
      autocompletedSuggestions
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    ).reduce((acc, [_, value]: [any, any]) => acc + value.length, 0);
    const allAutocompletedItemsCount =
      autocompletedSuggestionsCount + autocompletedResults.length;

    let handleOnSelectAutocomplete;
    if (onSelectAutocomplete) {
      handleOnSelectAutocomplete = (selection) => {
        onSelectAutocomplete(
          selection,
          {
            notifyAutocompleteSelected: this.handleNotifyAutocompleteSelected,
            completeSuggestion: this.completeSuggestion,
            autocompleteResults: this.props.autocompleteResults
          },
          this.defaultOnSelectAutocomplete
        );
      };
    }

    const viewProps = {
      allAutocompletedItemsCount: allAutocompletedItemsCount,
      autocompleteView,
      autocompleteResults: autocompleteResults,
      autocompleteSuggestions: autocompleteSuggestions,
      autocompletedResults: autocompletedResults,
      autocompletedSuggestions: autocompletedSuggestions,
      className,
      autocompletedSuggestionsCount: autocompletedSuggestionsCount,
      completeSuggestion: this.completeSuggestion,
      isFocused: isFocused,
      notifyAutocompleteSelected: this.handleNotifyAutocompleteSelected,
      onChange: (value) => this.handleChange(value),
      onSelectAutocomplete:
        handleOnSelectAutocomplete || this.defaultOnSelectAutocomplete,
      onSubmit: onSubmit
        ? (e) => {
            e.preventDefault();
            onSubmit(searchTerm);
          }
        : this.handleSubmit,
      useAutocomplete: useAutocomplete,
      value: searchTerm,
      inputProps: {
        onFocus: this.handleFocus,
        onBlur: this.handleBlur,
        ...inputProps
      },
      inputView,
      ...rest
    };

    return <View {...viewProps} />;
  }
}

export default withSearch<SearchBoxContainerProps, SearchBoxContainerContext>(
  ({
    autocompletedResults,
    autocompletedSuggestions,
    searchTerm,
    setSearchTerm,
    trackAutocompleteClickThrough,
    trackAutocompleteSuggestionClickThrough
  }) => ({
    autocompletedResults,
    autocompletedSuggestions,
    searchTerm,
    setSearchTerm,
    trackAutocompleteClickThrough,
    trackAutocompleteSuggestionClickThrough
  })
)(SearchBoxContainer);
