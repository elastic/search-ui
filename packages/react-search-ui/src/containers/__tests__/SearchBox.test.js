import React from "react";
import { SearchBoxContainer } from "../SearchBox";
import { shallow } from "enzyme";

const params = {
  autocompletedResults: [],
  autocompletedSuggestions: {},
  searchTerm: "test",
  setSearchTerm: jest.fn(),
  trackAutocompleteClickThrough: jest.fn()
};

beforeEach(() => {
  params.setSearchTerm = jest.fn();
  params.trackAutocompleteClickThrough = jest.fn();
});

it("supports a render prop", () => {
  // eslint-disable-next-line react/prop-types
  const render = ({ value }) => {
    return <div>{value}</div>;
  };
  const wrapper = shallow(<SearchBoxContainer {...params} view={render} />);
  expect(wrapper).toMatchSnapshot();
});

it("will keep focus prop in sync with view component", () => {
  let viewProps;

  shallow(
    <SearchBoxContainer {...params} view={props => (viewProps = props)} />
  );

  expect(viewProps.isFocused).toBe(false);
  viewProps.inputProps.onFocus();

  expect(viewProps.isFocused).toBe(true);

  viewProps.inputProps.onBlur();

  expect(viewProps.isFocused).toBe(false);
});

it("will pass autocompleteView prop through to the view", () => {
  let viewProps;
  const customAutocompleteView = () => {};

  shallow(
    <SearchBoxContainer
      {...params}
      autocompleteView={customAutocompleteView}
      view={props => (viewProps = props)}
    />
  );

  const { autocompleteView } = viewProps;
  expect(autocompleteView).toEqual(customAutocompleteView);
});

describe("autocompletedSuggestionsCount", () => {
  it("will calculate the total count of all suggestions", () => {
    let viewProps;

    shallow(
      <SearchBoxContainer
        {...params}
        autocompletedSuggestions={{
          documents: [
            { suggestion: "carlsbad" },
            { suggestion: "carlsbad caverns" },
            { suggestion: "carolina" }
          ],
          other: [],
          another: [{ suggestion: "carlsbad" }]
        }}
        view={props => (viewProps = props)}
      />
    );
    viewProps.onChange("new term");
    expect(viewProps.autocompletedSuggestionsCount).toBe(4);
  });

  it("will calculate the total count of all suggestions when there are none", () => {
    let viewProps;

    shallow(
      <SearchBoxContainer
        {...params}
        autocompletedSuggestions={{}}
        autocompleteSuggestions={false}
        view={props => (viewProps = props)}
      />
    );
    viewProps.onChange("new term");
    expect(viewProps.autocompletedSuggestionsCount).toBe(0);
  });
});

describe("allAutocompletedItemsCount", () => {
  it("will calculate the total count of all autocomplete items", () => {
    let viewProps;

    shallow(
      <SearchBoxContainer
        {...params}
        autocompletedResults={[{ id: { raw: 1 } }, { id: { raw: 2 } }]}
        autocompletedSuggestions={{
          documents: [
            { suggestion: "carlsbad" },
            { suggestion: "carlsbad caverns" },
            { suggestion: "carolina" }
          ],
          other: [],
          another: [{ suggestion: "carlsbad" }]
        }}
        view={props => (viewProps = props)}
      />
    );
    viewProps.onChange("new term");
    expect(viewProps.allAutocompletedItemsCount).toBe(6);
  });

  it("will calculate the total count of all autocomplete items when there are none", () => {
    let viewProps;

    shallow(
      <SearchBoxContainer
        {...params}
        autocompleteResults={false}
        autocompletedResults={[]}
        autocompleteSuggestions={false}
        autocompletedSuggestions={{}}
        view={props => (viewProps = props)}
      />
    );
    viewProps.onChange("new term");
    expect(viewProps.allAutocompletedItemsCount).toBe(0);
  });
});

describe("useAutocomplete", () => {
  it("will be true if autocompleteResults configuration has been provided", () => {
    let viewProps;

    shallow(
      <SearchBoxContainer
        {...params}
        autocompleteResults={{
          titleField: "title",
          urlField: "nps_link"
        }}
        view={props => (viewProps = props)}
      />
    );
    viewProps.onChange("new term");
    expect(viewProps.useAutocomplete).toBe(true);
  });

  it("will be false if no autocomplete config has been provided", () => {
    let viewProps;
    shallow(
      <SearchBoxContainer {...params} view={props => (viewProps = props)} />
    );
    viewProps.onChange("new term");
    expect(viewProps.useAutocomplete).toBe(false);
  });

  it("will be true if autocompleteSuggestions configuration has been provided", () => {
    let viewProps;

    shallow(
      <SearchBoxContainer
        {...params}
        autocompleteSuggestions={{
          documents: {
            sectionTitle: "Suggested Queries"
          }
        }}
        view={props => (viewProps = props)}
      />
    );
    viewProps.onChange("new term");
    expect(viewProps.useAutocomplete).toBe(true);
  });

  it("will be false if autocompleteMinimumCharacters is below threshold", () => {
    let viewProps;

    shallow(
      <SearchBoxContainer
        {...params}
        autocompleteMinimumCharacters={4}
        autocompleteSuggestions={{
          documents: {
            sectionTitle: "Suggested Queries"
          }
        }}
        view={props => (viewProps = props)}
      />
    );
    viewProps.onChange("new");
    expect(viewProps.useAutocomplete).toBe(true);
  });
});

describe("shouldClearFilters prop", () => {
  it("will be passed through to setSearchTerm on submit", () => {
    let viewProps;

    shallow(
      <SearchBoxContainer
        {...params}
        shouldClearFilters={false}
        view={props => (viewProps = props)}
      />
    );

    const { onSubmit } = viewProps;
    onSubmit({
      preventDefault: () => {}
    });
    const call = params.setSearchTerm.mock.calls[0];
    expect(call[1].shouldClearFilters).toEqual(false);
  });

  it("will be passed through to setSearchTerm on change", () => {
    let viewProps;

    shallow(
      <SearchBoxContainer
        {...params}
        shouldClearFilters={false}
        view={props => (viewProps = props)}
      />
    );

    const { onChange } = viewProps;
    onChange("new term");
    const call = params.setSearchTerm.mock.calls[0];
    expect(call[1].shouldClearFilters).toEqual(false);
  });

  it("will call setSearchTerm if no onSelectAutocomplete is specified and a suggestion is selected", () => {
    let viewProps;

    shallow(
      <SearchBoxContainer
        {...params}
        autocompleteResults={true}
        shouldClearFilters={false}
        view={props => (viewProps = props)}
      />
    );

    const { onSelectAutocomplete } = viewProps;
    onSelectAutocomplete({
      suggestion: "bird"
    });

    const call = params.setSearchTerm.mock.calls[0];
    expect(call[1].shouldClearFilters).toEqual(false);
  });
});

it("will call back to setSearchTerm with refresh: false when input is changed", () => {
  let viewProps;
  shallow(
    <SearchBoxContainer {...params} view={props => (viewProps = props)} />
  );

  expect(viewProps.value).toBe("test");

  viewProps.onChange("new term");

  const call = params.setSearchTerm.mock.calls[0];
  expect(call).toEqual([
    "new term",
    {
      refresh: false,
      autocompleteResults: false,
      autocompleteSuggestions: false,
      shouldClearFilters: true,
      autocompleteMinimumCharacters: 0
    }
  ]);
});

it("will call back to setSearchTerm with autocompleteMinimumCharacters setting", () => {
  let viewProps;
  shallow(
    <SearchBoxContainer
      {...params}
      autocompleteMinimumCharacters={3}
      view={props => (viewProps = props)}
    />
  );
  viewProps.onChange("new term");

  const call = params.setSearchTerm.mock.calls[0];
  expect(call).toEqual([
    "new term",
    {
      refresh: false,
      autocompleteResults: false,
      autocompleteSuggestions: false,
      shouldClearFilters: true,
      autocompleteMinimumCharacters: 3
    }
  ]);
});

it("will call back to setSearchTerm with refresh: true when input is changed and searchAsYouType is true", () => {
  let viewProps;
  shallow(
    <SearchBoxContainer
      {...params}
      searchAsYouType={true}
      view={props => (viewProps = props)}
    />
  );

  expect(viewProps.value).toBe("test");

  viewProps.onChange("new term");

  const call = params.setSearchTerm.mock.calls[0];
  expect(call).toEqual([
    "new term",
    {
      refresh: true,
      debounce: 200,
      autocompleteResults: false,
      autocompleteMinimumCharacters: 0,
      shouldClearFilters: true,
      autocompleteSuggestions: false
    }
  ]);
});

it("will call back to setSearchTerm with a specific debounce when input is changed and searchAsYouType is true and a debounce is provided", () => {
  let viewProps;
  shallow(
    <SearchBoxContainer
      {...params}
      searchAsYouType={true}
      debounceLength={500}
      view={props => (viewProps = props)}
    />
  );

  expect(viewProps.value).toBe("test");

  viewProps.onChange("new term");

  const call = params.setSearchTerm.mock.calls[0];
  expect(call).toEqual([
    "new term",
    {
      refresh: true,
      debounce: 500,
      autocompleteResults: false,
      autocompleteMinimumCharacters: 0,
      shouldClearFilters: true,
      autocompleteSuggestions: false
    }
  ]);
});

it("will call back to setSearchTerm with a specific debounce when input is changed and autocompleteResults is true and a debounce is provided", () => {
  let viewProps;
  shallow(
    <SearchBoxContainer
      {...params}
      autocompleteResults={{
        titleField: "title",
        urlField: "nps_link"
      }}
      debounceLength={500}
      view={props => (viewProps = props)}
    />
  );

  expect(viewProps.value).toBe("test");

  viewProps.onChange("new term");

  const call = params.setSearchTerm.mock.calls[0];
  expect(call).toEqual([
    "new term",
    {
      refresh: false,
      debounce: 500,
      autocompleteResults: true,
      autocompleteMinimumCharacters: 0,
      shouldClearFilters: true,
      autocompleteSuggestions: false
    }
  ]);
});

it("will call back to setSearchTerm with a specific debounce when input is changed and autocompleteSuggestions is true and a debounce is provided", () => {
  let viewProps;
  shallow(
    <SearchBoxContainer
      {...params}
      autocompleteSuggestions={true}
      debounceLength={500}
      view={props => (viewProps = props)}
    />
  );

  expect(viewProps.value).toBe("test");

  viewProps.onChange("new term");

  const call = params.setSearchTerm.mock.calls[0];
  expect(call).toEqual([
    "new term",
    {
      refresh: false,
      debounce: 500,
      autocompleteSuggestions: true,
      autocompleteMinimumCharacters: 0,
      shouldClearFilters: true,
      autocompleteResults: false
    }
  ]);
});

it("will call back setSearchTerm with refresh: true when form is submitted", () => {
  let viewProps;

  shallow(
    <SearchBoxContainer
      {...params}
      searchTerm="a term"
      view={props => (viewProps = props)}
    />
  );

  viewProps.onSubmit({
    preventDefault: () => {}
  });

  const call = params.setSearchTerm.mock.calls[0];
  expect(call).toEqual(["a term", { shouldClearFilters: true }]);
});

describe("onSelectAutocomplete", () => {
  it("will pass selection, option, default onSelectAutocomplete function to handler if provided", () => {
    let viewProps;

    let passedSelection;
    let passedNotifyAutocompleteSelected;
    let passedCompleteSuggestion;
    let passedAutocompleteResults;
    let passedDefaultOnSelectAutocomplete;

    const customOnSelectAutocomplete = (
      selection,
      { notifyAutocompleteSelected, completeSuggestion, autocompleteResults },
      defaultOnSelectAutocomplete
    ) => {
      passedSelection = selection;
      passedNotifyAutocompleteSelected = notifyAutocompleteSelected;
      passedCompleteSuggestion = completeSuggestion;
      passedAutocompleteResults = autocompleteResults;
      passedDefaultOnSelectAutocomplete = defaultOnSelectAutocomplete;
    };

    shallow(
      <SearchBoxContainer
        {...params}
        autocompleteResults={true}
        onSelectAutocomplete={customOnSelectAutocomplete}
        view={props => (viewProps = props)}
      />
    );
    const { onSelectAutocomplete } = viewProps;
    onSelectAutocomplete("bird");

    expect(passedSelection).toBe("bird");
    expect(passedNotifyAutocompleteSelected).toBeDefined();
    expect(passedCompleteSuggestion).toBeDefined();
    expect(passedAutocompleteResults).toBeDefined();
    expect(passedDefaultOnSelectAutocomplete).toBeDefined();
  });

  it("will call setSearchTerm if the default onSelectAutomplete is triggered with a suggestion", () => {
    let viewProps;

    shallow(
      <SearchBoxContainer
        {...params}
        autocompleteResults={true}
        view={props => (viewProps = props)}
      />
    );
    const { onSelectAutocomplete } = viewProps;
    onSelectAutocomplete({
      suggestion: "bird"
    });

    const call = params.setSearchTerm.mock.calls[0];
    expect(call[0]).toEqual("bird");
  });

  it("will do nothing if the default onSelectAutocomplete is triggered without a selection", () => {
    let viewProps;

    shallow(
      <SearchBoxContainer
        {...params}
        autocompleteResults={true}
        view={props => (viewProps = props)}
      />
    );
    const { onSelectAutocomplete } = viewProps;
    onSelectAutocomplete();

    expect(params.setSearchTerm).not.toHaveBeenCalled();
  });
});

describe("autocomplete clickthroughs", () => {
  it("will call back to trackAutocompleteClickThrough when an autocomplete item is selected in the view", () => {
    let viewProps;
    shallow(
      <SearchBoxContainer
        {...params}
        autocompleteResults={true}
        view={props => (viewProps = props)}
      />
    );
    const { notifyAutocompleteSelected } = viewProps;
    notifyAutocompleteSelected({
      id: { raw: "123" }
    });
    expect(params.trackAutocompleteClickThrough.mock.calls.length).toBe(1);
  });

  it("will not call back when shouldTrackClickThrough is false", () => {
    let viewProps;
    shallow(
      <SearchBoxContainer
        {...params}
        autocompleteResults={{
          shouldTrackClickThrough: false,
          titleField: "title",
          urlField: "nps_link"
        }}
        view={props => (viewProps = props)}
      />
    );
    const { notifyAutocompleteSelected } = viewProps;
    notifyAutocompleteSelected({
      id: { raw: "123" }
    });
    expect(params.trackAutocompleteClickThrough.mock.calls.length).toBe(0);
  });

  it("will not call back when the selected item is a suggestion", () => {
    let viewProps;
    shallow(
      <SearchBoxContainer
        {...params}
        autocompleteResults={true}
        view={props => (viewProps = props)}
      />
    );
    const { notifyAutocompleteSelected } = viewProps;
    notifyAutocompleteSelected({
      suggestion: "bike"
    });
    expect(params.trackAutocompleteClickThrough.mock.calls.length).toBe(0);
  });

  it("will pass through tags", () => {
    let viewProps;
    shallow(
      <SearchBoxContainer
        {...params}
        autocompleteResults={{
          clickThroughTags: ["whatever"],
          titleField: "title",
          urlField: "nps_link"
        }}
        view={props => (viewProps = props)}
      />
    );
    const { notifyAutocompleteSelected } = viewProps;
    notifyAutocompleteSelected({
      id: { raw: "123" }
    });
    const [
      documentId,
      tags
    ] = params.trackAutocompleteClickThrough.mock.calls[0];
    expect(documentId).toEqual("123");
    expect(tags).toEqual(["whatever"]);
  });
});

it("passes className through to the view", () => {
  let viewProps;
  const className = "test-class";
  shallow(
    <SearchBoxContainer
      {...params}
      className={className}
      view={props => (viewProps = props)}
    />
  );
  expect(viewProps.className).toEqual(className);
});

it("passes data-foo through to the view", () => {
  let viewProps;
  const data = "bar";
  shallow(
    <SearchBoxContainer
      {...params}
      data-foo={data}
      view={props => (viewProps = props)}
    />
  );
  expect(viewProps["data-foo"]).toEqual(data);
});
