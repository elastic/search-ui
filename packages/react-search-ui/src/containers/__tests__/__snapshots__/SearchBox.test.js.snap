// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<SearchBox
  autocompletedResults={Array []}
  autocompletedSuggestions={Object {}}
  inputProps={
    Object {
      "onBlur": [Function],
      "onFocus": [Function],
    }
  }
  isFocused={false}
  notifyAutocompleteSelected={[Function]}
  onChange={[Function]}
  onSubmit={[Function]}
  useAutocomplete={false}
  value="test"
/>
`;

exports[`supports a render prop 1`] = `
<div>
  test
</div>
`;
